include/charmlite/charmlite.hpp

```diff
@@ -1,21 +1,252 @@
+
+    // helper struct to erase type of combiners/callbacks
+    template <typename Message, template <class> class Function,
+        Function<Message> Fn, typename Enable = void>
+    struct function_wrapper_;
+
+    template <template <class> class Function, Function<message> Fn>
+    struct function_wrapper_<message, Function, Fn>
+    {
+        // no type erasure needed for the "base" case
+        static constexpr Function<message> fn(void)
+        {
+            return Fn;
+        }
+    };
+
+    template <typename Message, combiner_fn_t<Message> Fn>
+    struct function_wrapper_<Message, combiner_fn_t, Fn,
+        typename std::enable_if<!std::is_same<message, Message>::value>::type>
+    {
+        static message_ptr<> impl_(message_ptr<>&& lhs, message_ptr<>&& rhs)
+        {
+            // capture and retype pointers
+            auto* lhs_typed = static_cast<Message*>(lhs.release());
+            message_ptr<Message> lhs_owned(lhs_typed);
+            auto* rhs_typed = static_cast<Message*>(rhs.release());
+            message_ptr<Message> rhs_owned(rhs_typed);
+            // ( result should be implicitly castable to message )
+            return Fn(std::move(lhs_owned), std::move(rhs_owned));
+        }
+
+        static constexpr combiner_fn_t<message> fn(void)
+        {
+            return &(function_wrapper_<Message, combiner_fn_t, Fn>::impl_);
+        }
+    };
+
+    template <typename Message, callback_fn_t<Message> Fn>
+    struct function_wrapper_<Message, callback_fn_t, Fn,
+        typename std::enable_if<!std::is_same<message, Message>::value>::type>
+    {
+        static void impl_(message_ptr<>&& msg)
+        {
+            auto* typed = static_cast<Message*>(msg.release());
+            message_ptr<Message> owned(typed);
+            Fn(std::move(owned));
+        }
+
+        static constexpr callback_fn_t<message> fn(void)
+        {
+            return &(function_wrapper_<Message, callback_fn_t, Fn>::impl_);
+        }
+    };
+
+    template <typename Message, template <class> class Function,
+        Function<Message> Fn>
+    static std::size_t register_function_(std::vector<Function<message>>& table)
+    {
+        // get a type-erased version of the function
+        constexpr auto fn = function_wrapper_<Message, Function, Fn>::fn();
+        // then register it
+        auto id = table.size() + 1;
+        table.emplace_back(fn);
+        return id;
+    }
+
+    template <typename Message, combiner_fn_t<Message> Fn>
+    combiner_id_t combiner_helper_<Message, Fn>::id_ =
+        register_function_<Message, combiner_fn_t, Fn>(
+            CsvAccess(combiner_table_));
+
+    template <typename Message, callback_fn_t<Message> Fn>
+    callback_id_t callback_helper_<Message, Fn>::id_ =
+        register_function_<Message, callback_fn_t, Fn>(
+            CsvAccess(callback_table_));
+
+    inline completion* system_detector_(void)
```

Comment:
Now that encapsulation is working correctly, `system_detector_` should be moved into _core.cpp_.